### PR TITLE
Feat/lw 9027 cardano submit api

### PR DESCRIPTION
**Pull request by iccicci**
# Context

We need to implement tx submission through `cardano-submit-api`.

# Proposed Solution

- Implemented `TxSubmitApiProvider` in `cardano-services-client`.
- Implemented tx submission through `TxSubmitApiProvider` in `provider-server`.

# TODO

- A complete error handling.
- `TxSubmitApiProvider` unit tests.

### CODE REVIEW

File: compose/common.yml [hunk 1]
```diff
@@ -64,8 +64,10 @@ x-provider-server-environment: &provider-server-environment
   ENABLE_METRICS: ${ENABLE_METRICS:-false}
   EPOCH_POLL_INTERVAL: ${EPOCH_POLL_INTERVAL:-500}
   SERVICE_NAMES: ${SERVICE_NAMES:-asset,chain-history,handle,network-info,rewards,stake-pool,tx-submit,utxo}
+  SUBMIT_API_URL: ${SUBMIT_API_URL:-http://cardano-submit-api:8090/api/submit/tx}
```
Comment (rhyslbw) [inline compose/common.yml:67]:
```suggestion
  SUBMIT_API_URL: ${SUBMIT_API_URL:-http://cardano-submit-api:8090}
```
No need to expose the internal path of the service, just the root URL

File: packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+      // TODO improve (implement?) error handling
+      throw axios.isAxiosError(error)
+        ? new ProviderError(ProviderFailure.BadRequest, null, error.response?.data as string)
+        : new ProviderError(ProviderFailure.Unknown, error, 'submitting tx');
+    }
+  }
+
+  healthCheck() {
+    return Promise.resolve({ ok: true });
```
Comment (rhyslbw) [inline packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts:45]: I was going to suggest calling a path we know not to exist, and assert a 404, but it's more important to ship something that works ASAP. Maybe we should rethink if this method is even needed 🤔

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts (hunk 2 above)
Comment (iccicci) [reply packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts:45]: > I was going to suggest calling a path we know not to exist, and assert a 404
I thought the same, but I wasn't able to find a way to perform an ok dry call to get the `ok` status.

File: compose/common.yml (hunk 1 above)
Comment (iccicci) [reply compose/common.yml:67]: This is the `provider-server` configuration which contains the docker compose internal URL, nothing new is exposed.

Comment (rhyslbw) [reply compose/common.yml:67]: Right, but it's better to just configure the root URL, then add `/api/submit/tx` inside the client

File: packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts (hunk 2 above)
Comment (mkazlauskas) [reply packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts:45]: An improvement over what we currently have is mapping status of last submission: `2xx` or `4xx` is ok, everything else is not ok

Comment (mkazlauskas) [reply packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts:45]: I don't think it's meaningless: afaik client is not checking health before submitting, but health is used to track backend services

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
Nice work! Just a minor change request below, and also one of the commits scope is `2e2` (should be `e2e`)

File: packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  #path: string;
+
+  constructor(submitApiUrl: URL, logger: Logger) {
+    this.#axios = axios.create({ baseURL: submitApiUrl.origin });
+    this.#logger = logger;
+    this.#path = submitApiUrl.pathname;
+  }
+
+  async submitTx({ signedTransaction }: SubmitTxArgs) {
+    let txId = '' as Cardano.TransactionId;
```
[the remark this answers is not in the document]
Comment (iccicci) [reply packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts]: 10c97ca4c should solve this

File: packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts (hunk 2 above)
Comment (iccicci) [reply packages/cardano-services-client/src/TxSubmitProvider/TxSubmitApiProvider.ts]: b6dc40411 should solve this

File: compose/common.yml (hunk 1 above)
Comment (iccicci) [reply compose/common.yml]: b6dc40411 and 81d6db18f should solve this

**Review by mkazlauskas — APPROVED** [review]
Great work!

**Review by mirceahasegan — APPROVED** [review]
Fantastic work, @iccicci 🧙‍♂️ 🚀

**Review by rhyslbw — APPROVED** [review]
Great work @iccicci 🚀

**Pull request merged**